11_FFTBloom/app_resources/common.hlsl

```diff
@@ -0,0 +1,76 @@
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+
+#define CHANNELS 3
+
+using namespace nbl::hlsl;
+
+struct PushConstantData
+{
+	// After running FFT along a column, we want to store the result in column major order for coalesced writes, and similarly after running an FFT in row major order
+	// All FFTs that read from a buffer and write to a buffer (kernel second FFT, image second FFT + conv + IFFT) read in one order and write in another. 
+	// Since workgroups start and finish at arbitrary times it's not possible to pickup the data from a buffer in one layout and write it back in another layout due to
+	// possibility of writing over data that still hasn't been read by some workgroups.
+	uint64_t colMajorBufferAddress;
+	uint64_t rowMajorBufferAddress;
+	// To save some work, we don't mirror the image along both directions when doing the FFT. This means that when doing the FFT along the second axis, we do an FFT of length
+	// `RoundUpToPoT(dataElementCount + kernelPadding)` where `dataElementCount` is the actual length of the image along the second axis. We need it to keep track of the image's original dimension.
+	uint32_t dataElementCount;
+	float32_t2 kernelHalfPixelSize;
+	uint32_t numWorkgroupsLog2;
+};
+
+#ifdef __HLSL_VERSION
+
+#include "nbl/builtin/hlsl/complex.hlsl"
+#include "nbl/builtin/hlsl/bda/legacy_bda_accessor.hlsl"
+#include "nbl/builtin/hlsl/workgroup/fft.hlsl"
+
+[[vk::push_constant]] PushConstantData pushConstants;
+
+using namespace nbl::hlsl;
+
+template<uint16_t ElementsPerThread, uint16_t WorkgroupSize, typename Scalar>
+struct PreloadedAccessorBase {
+
+	void set(uint32_t idx, complex_t<Scalar> value)
+	{
+		preloaded[idx / WorkgroupSize] = value;
+	}
+
+	void get(uint32_t idx, NBL_REF_ARG(complex_t<Scalar>) value)
+	{
+		value = preloaded[idx / WorkgroupSize];
```

Comment:
in order to not allow someone to "fuck it up:" take `log2` of workgroup size as a parameter